1_2_bfs_2_3_tree/2_3_tree/main.go

```diff
@@ -0,0 +1,429 @@
+	for i := 0; i < n.size; i++ {
+		if n.key[i] == k {
+			return true
+		}
+	}
+	return false
+}
+
+func swap(x, y *int) {
+	*x, *y = *y, *x
+}
+
+func sort2(x, y *int) {
+	if *x > *y {
+		swap(x, y)
+	}
+}
+
+func sort3(x, y, z *int) {
+	if *x > *y {
+		swap(x, y)
+	}
+	if *x > *z {
+		swap(x, z)
+	}
+	if *y > *z {
+		swap(y, z)
+	}
+}
+
+func (n *Node) sort() {
+	switch n.size {
+	case 1:
+		return
+	case 2:
+		sort2(&n.key[0], &n.key[1])
+	case 3:
+		sort3(&n.key[0], &n.key[1], &n.key[2])
+	}
+}
+
+func (n *Node) insertToNode(k int) {
+	n.key[n.size] = k
+	n.size++
+	n.sort()
+}
+
+func (n *Node) removeFromNode(k int) {
+	if n.size >= 1 && n.key[0] == k {
+		n.key[0] = n.key[1]
+		n.key[1] = n.key[2]
+		n.size--
+	} else if n.size == 2 && n.key[1] == k {
+		n.key[1] = n.key[2]
+		n.size--
+	}
+}
+
+func (n *Node) becomeNode2(k int, first, second *Node) {
+	n.key[0] = k
+	n.first = first
+	n.second = second
+	n.third = nil
+	n.fourth = nil
+	n.parent = nil
+	n.size = 1
+}
+
+func (n *Node) isLeaf() bool {
+	return n.first == nil && n.second == nil && n.third == nil
+}
+
+func split(item *Node) *Node {
+	if item.size < 3 {
+		return item
+	}
+
+	x := &Node{key: [3]int{item.key[0]}, first: item.first, second: item.second, parent: item.parent}
+	y := &Node{key: [3]int{item.key[2]}, first: item.third, second: item.fourth, parent: item.parent}
+
+	if x.first != nil {
+		x.first.parent = x
+	}
+	if x.second != nil {
+		x.second.parent = x
+	}
+	if y.first != nil {
+		y.first.parent = y
+	}
+	if y.second != nil {
+		y.second.parent = y
+	}
+
+	if item.parent != nil {
+		item.parent.insertToNode(item.key[1])
+
+		if item.parent.first == item {
+			item.parent.first = nil
+		} else if item.parent.second == item {
+			item.parent.second = nil
+		} else if item.parent.third == item {
+			item.parent.third = nil
+		}
+
+		if item.parent.first == nil {
+			item.parent.fourth = item.parent.third
+			item.parent.third = item.parent.second
+			item.parent.second = y
+			item.parent.first = x
+		} else if item.parent.second == nil {
+			item.parent.fourth = item.parent.third
+			item.parent.third = y
+			item.parent.second = x
+		} else {
+			item.parent.fourth = y
+			item.parent.third = x
+		}
+
+		tmp := item.parent
+		// Here we would typically free the item, but Go has garbage collection
+		return tmp
+	} else {
+		x.parent = item
+		y.parent = item
+		item.becomeNode2(item.key[1], x, y)
+		return item
+	}
+}
+
+func insert(p *Node, k int) *Node {
+	if p == nil {
+		return NewNode(k) // Create the first 2-3 node (root) if the tree is empty
+	}
+
+	if p.isLeaf() {
+		p.insertToNode(k) // Insert the key into the leaf node
+	} else if k <= p.key[0] {
+		p.first = insert(p.first, k) // Recursively insert into the first child
+	} else if (p.size == 1) || (p.size == 2 && k <= p.key[1]) {
+		p.second = insert(p.second, k) // Insert into the second child
+	} else {
+		p.third = insert(p.third, k) // Insert into the third child
+	}
```

Comment:
вот тут у тебя багуля, ты присваиваешь потомку его родителя, из-за этого происходит зацикливание в дереве, чего допускать нельзя)
```suggestion
	} else if k <= p.key[0] {
		insert(p.first, k) // Recursively insert into the first child
	} else if (p.size == 1) || (p.size == 2 && k <= p.key[1]) {
		insert(p.second, k) // Insert into the second child
	} else {
		insert(p.third, k) // Insert into the third child
	}
```